### DISCUSSION TITLE
Maintaining an internal homebrew tap.

### DISCUSSION THREAD

**Discussion by jsrice7391** [Homebrew/discussions#3159] [Everyday usage]
Hello all,

I am trying to build an internal homebrew tap for my company. I am struggling to find any docs on how this would work.

Right now, all the API calls to GitHub within homebrew are made without an Authorization header, which prevents users from reaching anything with an API key.

 I have the knowledge to get this and add a “if env var is present, add a header to the GitHub API calls which would allow users to access private sources and taps. 
 
My question becomes can I add the logic to the GitHub API logic that allows private repo access? Or is there already a documented way to accomplish such a thing?

Thank you in advance.

**Comment by carlocab** [comment]
Doesn't setting `HOMEBREW_GITHUB_API_TOKEN` or using the `osxkeychain` credential helper work?

**Reply by jsrice7391** [reply]
I thought so too, but I ran this:

```bash
$ HOMEBREW_GITHUB_API_TOKEN="CONFIRMED_TOKEN" brew install my-tool --debug
```
All the external downloads work until it gets to download the source of the `url` parameter. 

```bash
==> Downloading https://api.github.com/repos/my-org/my-tool/releases/tags/0.20.2
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.4.5\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.4\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --head https://api.github.com/repos/indigo-ag/dev-cli/releases/tags/0.20.2
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.4.5\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.4\)\ curl/7.64.1 --header Accept-Language:\ en --fail --progress-bar --retry 3 --location --remote-time --output /Users/distiller/Library/Caches/Homebrew/downloads/185441a905102cf37b8e76ce6853d464dca7d15ac099ae4bf9634a2c1ebb02f7--0.20.2.incomplete https://api.github.com/repos/my-org/my-tool/releases/tags/0.20.2
#=#=#
curl: (22) The requested URL returned error: 404
Error: dev-cli-beta: Failed to download resource "my-tool"
Download failed: https://api.github.com/repos/my-org/my-repo/releases/tags/0.20.2
$ nano Formula/dev-cli-beta.rb
```

I figured I would be able to see some form of password or token being used in the payload for the internal repo, but it's not there.

```bash
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.4.5\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.4\)\ curl/7.64.1 --header Accept-Language:\ en --retry 3 --location --head https://api.github.com/repos/my-org/my-tool/releases/tags/0.20.2
/usr/local/Homebrew/Library/Homebrew/shims/shared/curl --disable --cookie /dev/null --globoff --show-error --user-agent Homebrew/3.4.5\ \(Macintosh\;\ Intel\ Mac\ OS\ X\ 11.4\)\ curl/7.64.1 --header Accept-Language:\ en --fail --progress-bar --retry 3 --location --remote-time --output /Users/distiller/Library/Caches/Homebrew/downloads/185441a905102cf37b8e76ce6853d464dca7d15ac099ae4bf9634a2c1ebb02f7--0.20.2.incomplete https://api.github.com/repos/my-org/my-tool/releases/tags/0.20.2
```

This leads me down to where most of these API calls are being made: 

https://github.com/Homebrew/brew/blob/058be73b72f143b01be84f959a8e6b57121eca4d/Library/Homebrew/utils/github.rb#L50

I do not see any `API.open_rest` invocations that set headers to get past the authentication.

I think there is a major use case here for users using homebrew for private taps. 

Would it be better to fork homebrew and add the ability for these headers there or I do not mind adding this feature?

**Reply by carlocab** [reply]
Setting `HOMEBREW_DOCKER_REGISTRY_TOKEN` might work. See https://github.com/Homebrew/brew/pull/12673. That code might have changed since, though, so you'll want to double-check `download_strategy.rb`.

**Reply by jsrice7391** [reply]
@carlocab got the same result with setting the `HOMEBREW_DOCKER_REGISTRY_TOKEN`.  It still does not add the header needed to the `curl` command. Could I ask why a Docker API key came to mind? Just interested in how one uses it on Docker.

**Comment by gwuah** [comment]
@jsrice7391 Check out this gist https://gist.github.com/minamijoyo/3d8aa79085369efb79964ba45e24bb0e
let me know if you need more help. I'm in the process of setting up the same thing for my company.
